### PR TITLE
Add record and switch camera features

### PR DESCRIPTION
**Pull request by belanagy**
Both front and back facing cameras are utilized
Video is recorded and saved on the device's internal storage
Storage path getter is prepared also to provide image directory for later use

### CODE REVIEW

File: app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -1,3 +1,6 @@
 <resources>
     <string name="app_name">SCCameraView</string>
+    <string name="main_switch_button">SWITCH</string>
```
Comment (WonderCsabo) [inline app/src/main/res/values/strings.xml]: Use lowercase for these strings, the style will uppercase them.

Comment (WonderCsabo) [inline app/src/main/res/values/strings.xml]: Use lowercase for these strings, the style will uppercase them.

File: app/src/main/res/values/styles.xml [hunk 2]
```diff
@@ -6,6 +6,11 @@
         <item name="colorPrimary">@color/colorPrimary</item>
         <item name="colorPrimaryDark">@color/colorPrimaryDark</item>
         <item name="colorAccent">@color/colorAccent</item>
+        <item name="android:windowNoTitle">true</item>
```
Comment (WonderCsabo) [inline app/src/main/res/values/styles.xml]: Instead extend a `NoActionBar` theme.

File: app/src/main/res/values/styles.xml [hunk 3]
```diff
@@ -6,6 +6,11 @@
...
+        <item name="android:windowActionBar">false</item>
+        <item name="windowNoTitle">true</item>
+        <item name="windowActionBar">false</item>
+        <item name="android:windowFullscreen">true</item>
```
Comment (WonderCsabo) [inline app/src/main/res/values/styles.xml]: I do not think we need full screen for this.

File: sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java [hunk 4]
```diff
@@ -53,20 +61,32 @@ public void setCalculatedHeight(int calculatedHeight) {
...
 
-    private class SurfaceTextureListener implements TextureView.SurfaceTextureListener {
+    public abstract void switchCamera();
+
+    private class SurfaceTextureListener implements TextureView.SurfaceTextureListener, ActivityCompat.OnRequestPermissionsResultCallback {
 
         @Override
         public void onSurfaceTextureAvailable(SurfaceTexture surface, int width, int height) {
-            openCamera();
+            if (!(ContextCompat.checkSelfPermission(getContext(), Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED &&
```
Comment (WonderCsabo) [inline sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java]: We do not need permission check in this library, this is the responsibility of the client.

File: sccameraview/src/main/java/com/example/sccameraview/Camera2View.java [hunk 5]
```diff
@@ -148,12 +153,18 @@ public void openCamera() {
         }
     }
 
-    private String getFrontFacingCameraId() throws CameraAccessException {
+    public void switchCamera() {
```
Comment (WonderCsabo) [inline sccameraview/src/main/java/com/example/sccameraview/Camera2View.java]: An `abstract` method should be for this in the base class, don't you think?

**Review by WonderCsabo — CHANGES_REQUESTED** [review]

File: sccameraview/src/main/java/com/example/sccameraview/Camera2View.java (hunk 5 above)
Comment (belanagy) [reply sccameraview/src/main/java/com/example/sccameraview/Camera2View.java]: There is, implemented in both classes, with different camera release methods

Comment (WonderCsabo) [reply sccameraview/src/main/java/com/example/sccameraview/Camera2View.java]: Then this method is missing the `@Override` annotation.

Comment (belanagy) [reply sccameraview/src/main/java/com/example/sccameraview/Camera2View.java]: Indeed! :)

File: app/src/main/java/io/supercharge/sccameraview/MainActivity.java [hunk 6]
```diff
@@ -1,13 +1,103 @@
...
+    private Button recordButton;
+    private Button stopButton;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
+
+        if (isEveryPermissionGranted()) {
+            bindViews();
```
Comment (WonderCsabo) [inline app/src/main/java/io/supercharge/sccameraview/MainActivity.java]: You can bind the views without permission. The only thing is when you open the camera, everything else can be written outside of the `if`.

File: sccameraview/build.gradle [hunk 7]
```diff
@@ -13,5 +13,5 @@ android {
 }
 
 dependencies {
-    compile "com.android.support:support-annotations:25.3.1"
+    compile "com.android.support:support-core-utils:25.3.1"
```
Comment (WonderCsabo) [inline sccameraview/build.gradle]: We do not need this.

File: sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java [hunk 8]
```diff
@@ -53,20 +60,24 @@ public void setCalculatedHeight(int calculatedHeight) {
...
         if (CamcorderProfile.hasProfile(cameraId, CamcorderProfile.QUALITY_480P)) {
             return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_480P);
         }
-
         return CamcorderProfile.get(cameraId, CamcorderProfile.QUALITY_LOW);
     }
 
     abstract void openCamera();
 
+    public abstract void switchCamera();
```
Comment (WonderCsabo) [inline sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java:73]: We should also have a setter or attribute, which sets the used camera up-front.

File: sccameraview/src/main/java/com/example/sccameraview/Camera2View.java [hunk 9]
```diff
@@ -355,6 +366,41 @@ private static Size chooseVideoSize(Size[] choices) {
         return choices[choices.length - 1];
     }
 
+    private File getOutputMediaFile(int type){
```
Comment (WonderCsabo) [inline sccameraview/src/main/java/com/example/sccameraview/Camera2View.java:369]: Why do we need this method exaclty? The client should be able to decide where to save the file.

**Review by WonderCsabo — CHANGES_REQUESTED** [review]

File: sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java (hunk 8 above)
Comment (belanagy) [reply sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java:73]: camera is opened based on frontFacingCameraActive attribute, which is set true in the constructor, is that a bad practice?

File: sccameraview/src/main/java/com/example/sccameraview/Camera2View.java (hunk 9 above)
Comment (belanagy) [reply sccameraview/src/main/java/com/example/sccameraview/Camera2View.java:369]: it is supposed to be a temporary solution to be able to test recording and taking a photo. File is saved based on which function was called

File: sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java (hunk 8 above)
Comment (WonderCsabo) [reply sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java:73]: It is okay, just add a setter to it, so if the client wants back-facing, currently he must open the normal camera and then switch to back. He should be able to just open the back camera in the first place, by setting that.

File: sccameraview/src/main/java/com/example/sccameraview/Camera2View.java (hunk 9 above)
Comment (WonderCsabo) [reply sccameraview/src/main/java/com/example/sccameraview/Camera2View.java:369]: So this method will be removed later? I am confused.

Comment (belanagy) [reply sccameraview/src/main/java/com/example/sccameraview/Camera2View.java:369]: Yes, I plan to do a refactor on this

Comment (WonderCsabo) [reply sccameraview/src/main/java/com/example/sccameraview/Camera2View.java:369]: Okay.

File: sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java (hunk 8 above)
Comment (belanagy) [reply sccameraview/src/main/java/com/example/sccameraview/BaseCameraView.java:73]: Done and updated

**Review by WonderCsabo — APPROVED** [review]

**Pull request merged**